### PR TITLE
Fixes #1220  -  Removes ZooKeeper watches on table configs when table is deleted.

### LINKED ISSUE #1220: Explore removing unused zookeeper watchers
For per table config, Accumulo creates zookeeper watches per table per tablet servers.  When a table is deleted these watches may be left behind consuming memory on the ZK servers.  Accumulo may be able to use the following to remove these watchers.

https://issues.apache.org/jira/browse/ZOOKEEPER-1829

### PR DESCRIPTION
**Pull request by jzgithub1**
This stops the number of watchers in Zookeeper from growing without a limit and ultimately crashing ZooKeeper.  You have to whitelist the wchp four letter command in ZooKeeper to verify watches are actually removed when tables are deleted.  You have to add this line to the zoo.cfg command to whitelist the four letter commands:  4lw.commands.whitelist=*  

This command will show you the table configurations in ZooKeeper that have watches on them:
'echo wchp | nc localhost 2181 | grep 'conf/table\.''

This pull request can't be done with the one for #1225.

### CODE REVIEW

**Comment by jzgithub1** [comment]
> The pull request does not appear to be doing what it purports to do in the comments/log message/subject. That is, it does not appear to remove watches... but does appear to avoid adding some possibly unnecessary ones on nodes that don't yet exist. I _think_ that works because the case of watching for new child node creation is covered by the fact that ZooCache clears out cached child nodes when the parent node watcher is triggered, so the parent node watcher is sufficient.
> 

@ctubbsii ,  I have verified with my Uno instance and the ZooKeeper four letter command 'wchp' that this pull request does indeed remove all of the ZooKeeper watches related to an individual table when that table is deleted.  I verified with the 'createtable' and 'deletetable' commands in the accumulo shell.  There is nothing academic about it.  If 'wchp' is giving me an accurate state of the watchers in ZooKeeper then this pull request is removing watches and associated session information when tables are deleted.  Our current code just leaves all the table configuration watches for a table to remain watched all of the time when a table is deleted even though the actual Zoonodes are gone.   I can demo this PR at our next Hackathon.  PR #1422  which implements your idea in #1225  works well too and eliminates all the watches for sure.   If someone considers that never removing the watches when tables are deleted to be the central problem of running out of memory in Zookeeper,  than we can solve the issue of crashing ZooKeepers with this PR.  That was the hypothesis  that @keith-turner put forth in #1220 .

**Comment by jzgithub1** [comment]
@ctubbsii Below is output you see when you run **master branch** then the  'wchp' command above after a createtable command and then after a deletetable command in the Uno shell.  




**This is what how Accumulo works now on Master branch**
**Show the watches on a table after 'createtable table1' in Uno shell**
jzeiberg@dragonfly:~/github/fluo-uno$ showtablewatchers | egrep '[2-5]/conf' 
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.groups.enabled
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.iterator.scan.vers.opt.maxVersions
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.constraint.1
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.balancer
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.iterator.majc.vers.opt.maxVersions
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.iterator.minc.vers
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.iterator.scan.vers
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.majc.compaction.strategy
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.classpath.context
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.replication
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.iterator.minc.vers.opt.maxVersions
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.iterator.majc.vers

**Show the watches on  the table1 that is now deleted after 'deletetable table1' in Uno shell**
**This should be empty**
**This PR makes this list empty**
jzeiberg@dragonfly:~/github/fluo-uno$ showtablewatchers | egrep '[2-5]/conf' 
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.groups.enabled
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.balancer
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.split.endrow.size.max
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.compaction.minor.logs.threshold
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.split.threshold
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.majc.compaction.strategy
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.classpath.context
/accumulo/a6711212-52fb-4d8a-b0d3-c560fac06c86/tables/2/conf/table.replication
`

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java [hunk 1]
```diff
@@ -435,6 +475,46 @@ public T retry() {
...
+      return watched;
+
+    Matcher configMatcher = TABLE_SETTING_CONFIG_PATTERN.matcher(zPath);
+
+    if (configMatcher.matches()) {
+      Stat stat;
+      try {
+        stat = zooKeeper.exists(zPath, false);
+      } catch (KeeperException.BadVersionException | KeeperException.NoNodeException e1) {
+        throw new ConcurrentModificationException();
```
[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: Thanks.  I will just re-throw the exception.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java [hunk 2]
```diff
@@ -435,6 +475,46 @@ public T retry() {
...
+      }
+
+      if (stat != null) {
+        watched = true;
+        synchronized (znodeExists) {
+          znodeExists.add(zPath);
+        }
```
[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: The java.util.concurrent package doesn’t even have a class called ConcurrentHashSet, but from JDK 8 onwards, you can use newly added keySet(default value) and newKeySet() method to create a ConcurrentHashSet backed by ConcurrentHashMap.  I was originally using a HashMap that mapped zpaths to Boolean (always true).  @ivakegg correctly pointed out I should just be using a Set<> object.   I could use a ConcurrentHashMap<String, Boolean> I guess too.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java [hunk 3]
```diff
@@ -155,8 +167,22 @@ public void process(WatchedEvent event) {
...
 
       switch (event.getType()) {
-        case NodeDataChanged:
         case NodeChildrenChanged:
+          if (event.getPath().endsWith("conf")) {
+            Matcher confDirMatcher = TABLE_CONFIG_DIR_PATTERN.matcher(event.getPath());
+            if (confDirMatcher.matches()) {
+              try {
+                clear(event.getPath());
+                getZooKeeper().exists(event.getPath(), watcher);
```
[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: I would contend that it sets a data watch.  But the main idea here is that this event is triggered on a 'conf' directory in the first place and if  the watch on that znode is not reset it will not be triggered again which is why @ivakegg  and @keith-turner may have seen table config persistence anomalies over time.   Accumulo sets the ZCacheWatcher watcher member object in ZooCache to watch on the 'conf' directories outside of this case statement.  All I am doing here is actually taking a logical, reasonable and productive action when that occurs.  On master branch all that happens is that the value (null) associated with the 'conf' directory is deleted from ZooCache.  By doing that we are totally missing the opportunity to 'refresh' the configs for that table.  In the event that for some reason there is no watch yet on a table config then we will not get the event that it has been added or changed.

So to recap.  I am just taking a course of action here that I think is required and should be done and resetting the watch on that zoopath for the tables conf directory so the event will be retriggered if a child changes again.  Remember.  We are not setting the first watch here.  Its being done elsewhere for a good reason and we just drop the resulting event on the floor IMHO.

File: server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java [hunk 4]
```diff
@@ -250,6 +250,10 @@ public void removeTable(TableId tableId) throws KeeperException, InterruptedExce
       tableStateCache.remove(tableId);
       zoo.recursiveDelete(zkRoot + Constants.ZTABLES + "/" + tableId + Constants.ZTABLE_STATE,
           NodeMissingPolicy.SKIP);
+      String tableConfigPath = zkRoot + Constants.ZTABLES + "/" + tableId + Constants.ZTABLE_CONF;
+      // The line below removes all the watches on the table configuration items from the Zookeeper
+      // server
+      zoo.getZooKeeper().getChildren(tableConfigPath, false);
```
[the remark this answers is not in the document]
Comment (jzgithub1) [reply server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java]: When I started working this issue a few weeks back this code had an effect of removing watchers but the subsequent changes in the ZooCache seems to have made it unneeded.  Thanks for motivating me to retry the solution without this modification.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: According to the zookeeper documentation, it will set a data watch.  A quick look through the code it looks like the data is for the `...\conf\` node is initialized with `new byte[0]` and I don't see it set to anything else.  Is there somewhere that I'm missing that sets / uses the data directly on the `...\conf` node?  Or, is the `...\conf` node a parent node for the rest of the configuration?  If the `...\conf` node exists solely so that child nodes can be created / contained, what value is there in setting a data watch for a change?

[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: What I think @EdColeman is that the intention was to do what I am trying to do now.  We can't just do nothing when this event occurs.  It is a great opportunity to refresh the table configs for an individual table if the NodeChildrenChanged event for the table 'conf' directories can always reliably be captured.  Thanks for looking into the documentation.

[the remark this answers is not in the document]
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: To (hopefully) clarify my basic question - where in the code does something cause a data watch event to be triggered on the `...\conf` node?  What data is stored on the node, what uses it, and where is it modified after the initial creation?

[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: When some function like TablePropUtil.setTableProerty or TablePropUtil.deleteTableProperty is called ( or ZooUtil.putData during cloning) on a table config zoopath.  ZooKeeper will send a watch event of type NodeChildrenChange for the 'conf' zoopath to watcher in ZooCache and it will be processed in ZCacheWatcher.process().   Watchers are one-time triggers and need to be reset.  That is what this PR is doing now.

[the remark this answers is not in the document]
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: Right now, I'm not asking about child change events.  For the `exists(path, watch)` to fire a data change event, something needs to put data to that node - child change events are a different situation and are not going to be updated with the data watch placed in a zookeeper exists call.

[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: So what you are saying is that I need to call getZookeeper().getChildren(zpath, wathcher) to get the NodeChildrenChanged event when the children change.  I will give that a try @EdColeman .  Thanks for the information.

[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: @EdColeman .  What I am seeing during run time is that when a new table config parameter is set on a table (in the Accumulo shell) the NodeChildrenChanged watch event for the parameters parent 'conf' directory gets triggered for 'exist watchers'.  Setting a watch on the new table config parameter can be done in the case statement too along with resetting the watcher on the new child nodes as you mentioned above.   

This part of the PR really isn't the main part.  It really just looks like the safe and conservative thing to do. 

The main idea of the PR is to not create watches on znodes that don't exist yet and may never come into existence.  Only create watches on table configs actually in ZooKeeper.  By doing this the delete operation on the table actually deletes all the table configs zoonodes and the corresponding watches go away also.  This keeps ZooKeeper from expanding the watcher maps in the ZooKeeper class from growing to the point of making ZooKeeper run out of memory and crash.  

If this modification to the case statement is considered undesirable and without any merit then I can remove it I am pretty sure.  I just think it is a good way to ensure the cache is consistent with what is actually in ZooKeeper.

[the remark this answers is not in the document]
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: At this point, I'm trying to make sure that my understanding of zookeeper and my reading of the zookeeper documentation is correct so that these changes can be evaluated for correctness  
The original question that was asked is "if the watch should be added with an exits() or a getChildren() call". You maintain that this is the safe and conservative thing to do.  With my understanding of zookeeper, the watcher set by exists in this case is unnecessary.

Specifically, looking at [zookeeper user's guide - semantics of watches](http://zookeeper.apache.org/doc/r3.4.14/zookeeperProgrammers.html#ch_zkWatches) 

> - Created event: Enabled with a call to exists.
> - Deleted event: Enabled with a call to exists, getData, and getChildren.
> - Changed event: Enabled with a call to exists and getData.
> - Child event: Enabled with a call to getChildren.

This leads me to believe that  child events (node created, deleted, modified) are not enabled / called with an exists() call.  Using zookeeper directly, I performed the following:
```
created node /a
call exists(/a, true)  - this sets the exists watcher.
created node /a/b    (the exists watcher was not triggered)
call getChildren(/a, true) - this sets a second watcher for child events.
created node /a/c  - the child event watcher is triggered.
created node /a/d - no watchers triggered (watcher was not reset after /a/c create event)
modified the data on /a - the data watcher set with exists is triggered - also showing that it was not cleared with the child modification
```
This shows me that the exists() watcher is not triggered / cleared on child events - it is triggered when the data of node /a is modified.

You state that you see the exits watcher being triggered.  To me that implies something changed the data on the node.  Where in the code (other than initialization) is something changing the data?  Where in the code does anything read the data from the config node? (that would imply that it is being set / used). All I see is it being set at initialization (and on clone) to new byte[0].  Or, can you demonstrate a zookeeper call sequence that replicates what you are seeing?

My suspicion is the NodeChildrenChanged is being set with a call to getChildren() somewhere and that is what is being triggered, not with an exists() call.

I'm working an IT example to see if that adds clarity to zookeeper usage in the system beyond a toy zookeeper example directly.

[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: I am doing my testing and evaluation inside of the Accumulo shell in an UNO instance.  I tail my Accumulo logs with this command looking for my INFO debug statement:
tail -f install/logs/accumulo/* | grep 'NodeChildrenChanged: resetting watcher for'

Adding or deleting a table config will trigger the NodeChildrenChanged event to the 'conf' znode.  Do it in the Accumulo shell.  Try the steps below.

I repeatedly hit the NodeChildrenChanged case when I add a table config parameter to a table or delete a config parameter with 'config -t table1 -s table.bloom.enabled=true' or any other config that is not a default parameter.  Like I said,  I do this all in the Accumulo shell.  I use the zkCli.sh script to inspect the cversion and dataVersion fields with commands like 'get /accumulo/a328a8a6-78d6-4725-9bfc-80e5ec3a8639/tables/2/conf'. The cversion field changes the dataVersion doesn't.  I am reentering the NodeChildrenChanged case statement over and over because a reset the exists watcher on the node.  Watchers are one-time triggers and you have to reset them.  We were not doing that before.  Like I said before,  these lines of code are not central to this PR.  They just might help out though.  I ran the code with getZookeeper().get Children( zpath, watcher) to.  That did not harm anything.

The real evaluation of this PR is just verifying the removal of watches with the 'whcp' command I listed above - IMHO.  I can roll back to the old switch statement if these lines are causing confusion and they can be evaluated in the future.  I like them though.

We can do some pair programming at the Hack session next Tuesday and I can show you.

File: server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java]: I think that effect would have been seen because the watch had been triggered, and this code requests the children without re-setting that watch. However, if you don't care about the return value, simply not retrieving the children at all would have the same effect as calling `getchildren` with `false` for the watch and throwing away the return value. But, omitting the line does less work with the same effect.

The important thing to note is that watches will be removed over time if they are triggered and not re-set, because watches trigger at most once, and then get removed. If we don't re-set them, it has the same effect as it being deleted.

[the remark this answers is not in the document]
Comment (jzgithub1) [reply server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java]: Yeah.  But if they never get triggered they will never get deleted.  When a watch is created on a znode that is never created the watch stays around forever on ZooKeeper.  Also, Zookeeper keeps a session string for each service that has an association to a watch.  So, if you have a thousand node cluster, memory is taken to keep track of each services association to a ZooNode that doesn't exist and won't exist again.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: The most important thing I think here is that when the watch is re-set here, it re-sets the same watches that were triggered. If `exists` does that, that's fine. But, since we'd be using the parent node to watch for changes in the children (instead of watching directly for non-existent child node changes), it might need to use `getChildren` to ensure the watch is set to look for child node changes.

Much of this complexity goes away entirely if we were to transition from storing the config as separate child nodes of `conf/` and serialize the configuration directly into the data of `conf/` as a JSON blob. I think that is probably a more worthy pursuit than trying to fix this in ZooCache. I'm going to look into doing that in a separate PR. I've looked at it in the past, and I think I have a good idea of what needs to be done. It should be easier now than when I looked at it in the past, since observers were removed and the upgrade code was improved.

**Comment by ctubbsii** [comment]
> I have verified with my Uno instance and the ZooKeeper four letter command 'wchp' that this pull request does indeed remove all of the ZooKeeper watches related to an individual table when that table is deleted.

Because of https://issues.apache.org/jira/browse/ZOOKEEPER-1274, that command doesn't give a full picture of the watches (it doesn't show child watches), which concerns me a bit.

> @ctubbsii Below is output you see when you run **master branch** then the 'wchp' command shown below after a createtable command and then after a deletetable command in the Uno shell.

I see that it does have fewer data watches, and have no doubt of this. My main concern is that the way we use ZooCache, these watches might be important and may have negative consequences if removed. Although it fixes the problem of leaving watches around when the table is deleted, it may make it so ZooCache misses when a new node is created. This is somewhat a drop in the bucket, though. The number of watches is still `O(m*n*p)`, albeit with a slightly smaller `p`. I think this scale problem and the watches on non-existing nodes issue are both resolved by a simpler solution to consolidate configs into a single node, which (as I said above), I'm going to try to look into, picking up my previous investigation into that.

File: server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java]: Yes, I understand, but the context here is which kind of watcher do we make sure is put on the `conf/` node, not the child nodes.

**Comment by EdColeman** [comment]
Another thing that I've confirmed - calling exists(path, false) does not remove previous watchers - it just prevents adding a new one.

Using zookeeper, I called:
```
zoo.exists("/a", existsWatcher);
zoo.exists("/a", false);
zoo.exists("/a", null);
```
Still results in the watcher being triggered on a data change event.

File: server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jzgithub1) [reply server/base/src/main/java/org/apache/accumulo/server/tables/TableManager.java]: We set a data watch on the 'conf' node.  From the ZooKeeper webpage linked above:
>The data for which the watch was set This refers to the different ways a node can change. It helps to think of ZooKeeper as maintaining two lists of watches: data watches and child watches. getData() and exists() set data watches. getChildren() sets child watches. Alternatively, it may help to think of watches being set according to the kind of data returned. getData() and exists() return information about the data of the node, whereas getChildren() returns a list of children. Thus, setData() will trigger data watches for the znode being set (assuming the set is successful). A successful create() will trigger a data watch for the znode being created and a child watch for the parent znode. A successful delete() will trigger both a data watch and a child watch (since there can be no more children) for a znode being deleted as well as a child watch for the parent znode.

I took this line of code out anyways and pushed the change.  This solution does not require that line of code.

File: core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jzgithub1) [reply core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java]: The watches on the children nodes of the 'conf' directories will be set as data watches the first time that ZooCache.get is called on the individual zoopaths of those configs.  All of the table configurations will get watched eventually like before if they really exist in zookeeper.  I agree that calling getChildren seems like a wise idea but all you are doing is putting a watch on a new child that is not in the ZooCache yet anyways and will only get into it when ZooCache.get is finally called on that (child) table configurations zoopath.   @keith-turner  made a great point to me that we should be limiting calls to zookeeper inside of the process method.

**Comment by EdColeman** [comment]
I have a partial IT test that sets and checks watchers - https://github.com/EdColeman/accumulo/tree/zk_watcher_it.  There are two "tests" on for setting system props and one (in progress) for a table prop.  I still want to look at delete ops on a table to see if that is triggering an exists event, and there may be some formatting issues, but may be enough to look at until I revisit tomorrow.

**Comment by jzgithub1** [comment]
@EdColeman , thanks a lot.

**Comment by ctubbsii** [comment]
This is superseded by #1454
